src/Adapter/Order/CommandHandler/DeleteCartRuleFromOrderHandler.php

```diff
@@ -85,12 +100,74 @@ public function handle(DeleteCartRuleFromOrderCommand $command)
 
         try {
             // Delete Order Cart Rule and update Order
+            $cartRule = new CartRule($orderCartRule->id_cart_rule);
             $orderCartRule->softDelete();
             $cart->removeCartRule($orderCartRule->id_cart_rule);
 
-            $this->orderAmountUpdater->update($order, $cart, $orderCartRule->id_order_invoice);
+            // If cart rule was a gift product we must update an OrderDetail manually
+            $giftOrderDetail = $this->getGiftOrderDetail($order, $cartRule);
+            if (null !== $giftOrderDetail) {
+                $newQuantity = ((int) $giftOrderDetail->product_quantity) - 1;
+
+                /*
+                 * Note: we are lucky the stock updates happens smoothly, it was not re-injected by removing the CartRule
+                 * because Cart doesn't update the stock but it will while we update the OrderDetail Lucky
+                 * for us it's the same difference of 1 so it's all accurate in the end
+                 */
+                // This calls the OrderAmountUpdater internally so no need to perform both calls
+                $this->orderProductQuantityUpdater->update(
+                    $order,
+                    $giftOrderDetail,
+                    $newQuantity,
+                    $giftOrderDetail->id_order_invoice != 0 ? new OrderInvoice($giftOrderDetail->id_order_invoice) : null,
+                    false
+                );
+            } else {
+                $this->orderAmountUpdater->update($order, $cart, $orderCartRule->id_order_invoice);
+            }
         } finally {
             $this->contextStateManager->restoreContext();
         }
     }
+
+    /**
+     * @param Order $order
+     * @param CartRule $cartRule
+     *
+     * @return OrderDetail|null
+     */
+    private function getGiftOrderDetail(Order $order, CartRule $cartRule): ?OrderDetail
+    {
+        $productId = (int) $cartRule->gift_product;
+        $productAttributeId = (int) $cartRule->gift_product_attribute;
+
+        // First filter OrderDetails matching the gift
+        $giftOrderDetails = [];
+        foreach ($order->getOrderDetailList() as $orderDetail) {
+            if ((int) $orderDetail['product_id'] !== $productId || (int) $orderDetail['product_attribute_id'] !== $productAttributeId) {
+                continue;
+            }
+            $giftOrderDetails[] = $orderDetail;
+        }
+
+        if (empty($giftOrderDetails)) {
+            return null;
+        }
```

Comment:
I don't understand why you collect all if you only want the first one whose product_quantity is > 1.
You could do it here and get rid of the rest of the method:

```suggestion
        $fallbackOrderDetailId = null;
        $giftOrderDetailId = null;
        foreach ($order->getOrderDetailList() as $orderDetail) {
            if ((int) $orderDetail['product_id'] !== $productId || (int) $orderDetail['product_attribute_id'] !== $productAttributeId) {
                continue;
            }
            
            // We try to find a row with at least 2 items
            if ($orderDetail['product_quantity'] > 1) {
                $giftOrderDetailId = $orderDetail['id_order_detail'];
                return new OrderDetail($giftOrderDetailId);
            }

			// keep the first one for fallback
            $fallbackOrderDetailId ??= $orderDetail['id_order_detail'];
        }

        return (null === $fallbackOrderDetailId) ? null : new OrderDetail($fallbackOrderDetailId);
```

This may need some adjustments if `id_order_detail` can be null.

Reply:
Yep, that works too indeed!

Reply:
`id_order_detail` can not be null, the array is a result from a query that SELECT * from order_detail
If this table is inconsistent there are some much bigger issues that shouldn't be related to this handler ^^